### DISCUSSION TITLE
Define key encoding/decoding strategy when using Codable with Firestore

### DISCUSSION THREAD

**Discussion by joshbruhin** [firebase/firebase-ios-sdk#10344] [Ideas]
When using the FirebaseFirestoreSwift API to easily encode and decode Codable objects when saving and retrieving documents from Firestore, is it possible to set the `.keyEncodingStrategy` and `.keyDecodingStrategy` to be used by the Firestore encoder/decoder?

**Comment by mortenbekditlevsen** [comment]
That ought to be possible with the newly released v10 of the sdk.

**Reply by paulb777** [reply]
Thanks to @mortenbekditlevsen both for answering here and doing the implementation! 🎊

**Reply by andrewheard** [reply]
As simple example for anyone coming across this, if your keys were in snake_case you could read or write the following document:
```
let userRef = Firestore.firestore().collection("users").document()
```

**Read**
```
let decoder = Firestore.Decoder()
decoder.keyDecodingStrategy = .convertFromSnakeCase

let user: User? = try await userRef.getDocument(as: User.self, decoder: decoder)
```

**Write**
```
let encoder = Firestore.Encoder()
encoder.keyEncodingStrategy = .convertToSnakeCase

try userRef.setData(from: user, encoder: encoder)
```

Big thanks to @mortenbekditlevsen for this feature!

**Reply by KhaledShehadeh** [reply]
Just a quick note.
Do not use a same shared instance of decoder and encoder. This will result in a thread race.
